// @linaria/babel-preset/lib/plugins/babel-transform.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = babelTransform;
var _logger = require("@linaria/logger");
var _utils = require("@linaria/utils");
var _cache = require("../cache");
var _transform = require("../transform");
var _loadLinariaOptions = _interopRequireDefault(require("../transform/helpers/loadLinariaOptions"));
var _collector = require("./collector");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function babelTransform(babel, options) {
  const cache = new _cache.TransformCacheCollection();
  return {
    name: '@linaria/babel/babel-transform',
    pre(file) {
      var _file$opts$root, _file$opts$inputSourc;
      // eslint-disable-next-line require-yield
      function* collect() {
        const {
          valueCache
        } = this.data;
        const {
          loadedAndParsed
        } = this.entrypoint;
        const {
          pluginOptions
        } = this.services.options;
        if (loadedAndParsed.evaluator === 'ignored') {
          throw new Error('entrypoint was ignored');
        }
        (0, _collector.collector)(file, pluginOptions, valueCache);
        return {
          ast: loadedAndParsed.ast,
          code: loadedAndParsed.code
        };
      }
      (0, _logger.debug)('babel-transform:start', file.opts.filename);
      const pluginOptions = (0, _loadLinariaOptions.default)(options);
      (0, _transform.transformSync)({
        babel,
        cache,
        options: {
          filename: file.opts.filename,
          root: (_file$opts$root = file.opts.root) !== null && _file$opts$root !== void 0 ? _file$opts$root : undefined,
          inputSourceMap: (_file$opts$inputSourc = file.opts.inputSourceMap) !== null && _file$opts$inputSourc !== void 0 ? _file$opts$inputSourc : undefined,
          pluginOptions
        }
      }, file.code, _utils.syncResolve, {
        collect
      });
    },
    visitor: {},
    post(file) {
      (0, _logger.debug)('babel-transform:end', file.opts.filename);
    }
  };
}
